### ISSUE TITLE
Add support for time limits for Web API permissions in brave-core

### ISSUE THREAD

**Issue by pes10k** [brave/brave-browser#14126]
Currently, when requesting a permission, Chromium allows users to say yes or no, w/o an intermediate option.

For example, i can give google maps access to my location forever (or deny), or i can give the zoom web client access to my camera and mic forever (or deny), but there is no option to say "you can use my camera until i close the page" or "you can use my location for 24 hours".

## Options
This issue is to create the ability to grant a page access to a permissioned API for the following options:
1. Grant permission, forever (currently available)
2. No permission, forever (currently available)
3. Permission for this site session (i.e. until there are no more top-level documents for the eTLD+1, to align with ephemeral site storage)
4. Permission for `arbitrary amount of time`

For 4, we likely wouldn't want to expose this option to users (I imagine we'd want so preset "one week" or similar options), but the implementation should allow for arbitrary lengths.

## Current UI
Currently Brave uses the upstream Chromium options (and UI), which give the user only a permanent  Yes or No option.
<img width="367" alt="Screen Shot 2021-02-11 at 8 19 15 PM" src="https://user-images.githubusercontent.com/76526/107729770-9da2df80-6ca6-11eb-9fb2-0db49426bc2b.png">

### Fiirefox
<img width="399" alt="Screen Shot 2021-02-11 at 8 10 07 PM" src="https://user-images.githubusercontent.com/76526/107729268-3afd1400-6ca5-11eb-8c76-b0205a6be58e.png">
<img width="372" alt="Screen Shot 2021-02-11 at 8 12 16 PM" src="https://user-images.githubusercontent.com/76526/107729464-dbebcf00-6ca5-11eb-8e94-b3968ba67393.png">
Firefox currently has a version of this for for some permissions. By default, "allow" and "deny" apply for the page, and "Remember this decision" makes the choice permanent.

### Safari
Safari similarly by default grants permission for the page, with options to make the choice permeant.
<img width="290" alt="Screen Shot 2021-02-11 at 8 13 49 PM" src="https://user-images.githubusercontent.com/76526/107729492-f32abc80-6ca5-11eb-97c7-caba6709ceca.png">
<img width="273" alt="Screen Shot 2021-02-11 at 8 14 41 PM" src="https://user-images.githubusercontent.com/76526/107729493-f3c35300-6ca5-11eb-8121-46fc1d825526.png">

## Out-of-Scope and Other Considerations
- This issue does not cover the UI to expose these new capabilities to users.  UI work is covered in issues https://github.com/brave/brave-browser/issues/14127 (Desktop) and https://github.com/brave/brave-browser/issues/14128 (Android)
- It would be ideal to extend these lifetime'd permissions to extension permissions, but that is separate, come later work
- I expect it may be difficult to surface the lifetime'd permissions in the chrome content settings UI. That would also be ideal, but is also saved for follow up work.
- This issue currently suggests adding "for site session" (in the [ephemeral site storage](https://brave.com/privacy-updates-7/) sense).  If this ends up being significantly more difficult than the "page session" options Firefox and Safari use, lets change this issue to be "page session" too

**Comment by goodov** [comment]
Did some research, a possible solution looks like this:
1. Create additional ```KeyedService``` to track expiring permissions and revoke them when it's required. The proper place is ```components/permissions```.
2. Reach this service and pass all data when we got the decision from a user. It can be done in the implementation of ```PermissionPrompt::Delegate```, currently it is implemented in ```PermissionRequestManager``` (a tab helper). To pass the actual value from UI, we can extend ```PermissionPrompt::Delegate::Accept``` or add another method and call it right (or before) ```Accept``` with the desired amount of time.
   It is also possible to extend ```PermissionRequestImpl``` by adding ```BrowserContext``` and access the service directly in case ```PermissionRequestManager``` will be hard to wire.

❔ It looks like a request bubble can show many permission requests at once. The current delegate implementation assumes we accept all or decline all. The default timer implementation will also be applied to all listed permissions. Can we mention that in the issue description? I guess we can technically support different timers for different permissions requested simultaneously in the one UI bubble, but is it actually needed?

I haven't look at ephemeral storage-like behavior yet, but I assume it should not be that hard to implement knowing that the ephemeral storage feature already has some working machinery for a tab tracking which can be connected with this feature.

**Comment by pes10k** [comment]
That all sounds great @goodov! Also wondering if your digging gave you a sense of how difficult it would be to partition permissions (e.g., whether google maps embedded on site A is independent of whether google maps embedded on site B, both of which are independent of what permissions https://www.google.com/maps has, etc).

Its possible upstream is already considering this (as part of the overall NetworkIsolationKey work) but if they haven't / aren't planning to, we should at least.

If that would be "easy" to do as part of this issue, I'll add and update the description above. But if it seems like a significant amount of additional work, i'll create another issue for it, and we can sequence however you both think would be best.

**Comment by iefremov** [comment]
@pes10k Let's do step by step, having an additional issue would be fine

**Comment by goodov** [comment]
Extension API permissions logic uses another mechanism, not connected with ```PermissionManager```, so it will not automagically work when we implement the Web API part. Additional implementation for the extension API is required.

Entry point for extension API permissions is ```PermissionsRequestFunction::Run```, then the flow goes into ```PermissionsUpdater``` class. It looks like we can use this class directly in our backend to reset permissions when it's required.

**Comment by pes10k** [comment]
@goodov would your current approach also work for an "allow until end of session" option?

**Comment by goodov** [comment]
@pes10k There are currently two approaches now that I'm considering.

1. Use existing Chromium logic, it's a good approach, because it minimizes overall logic duplication between Brave and Chromium, but it will require us to make a CL to straighten the underlying interfaces in Chromium codebase (otherwise the conflict score will be super high), and possibly extend their "per session" logic so it will conform to our eTLD+1 idea instead of theirs Origin-based implementation.
2. Create a Brave-specific service on our side and do all permission revoke steps ourselves. This way we can apply any rules we want and be not restricted by Chromium logic.

The Chromium-based option is very appealing, but it can take a lot of time, and the final result may not be as flexible as we want to, so we discussed it with Ivan and decided to implement Brave-specific variant with possible Chromium reintegration in future.

> would your current approach also work for an "allow until end of session" option?

the current approach with our service will allow me to connect it with ```TLDEphemeralLifetime``` and handle "per session" lifetime without noticeable issues.

**Comment by goodov** [comment]
Here is a discussion with Chromium folks about improving their permission constraints implementation so we can use it directly at some point. Looks like they're happy to accept my recommended changes and we can actually do this. Just need some time to implement it and made a CL.

https://bugs.chromium.org/p/chromium/issues/detail?id=1147918#c4

**Comment by pes10k** [comment]
That sounds terrific @goodov! Would we need to wait for upstream before we can enable for Brave users then?  Or can we enable in Brave while the upstream work is being sorted out?

**Comment by goodov** [comment]
We will enable our implementation first (currently in review) and will migrate to upstream implementation when it will be ready (may take few months to get required Chromium changes in our codebase).

**Comment by goodov** [comment]
There are a list of some permission requests that are not handled right now. Some of them use own storage (not ```HostContentSettingsMap```), some of them use custom ```HostContentSettingsMap``` decision set logic, some of them are even browser-wide:
1. ```WidevinePermissionRequest``` - Widevine enabler for the whole browser. Doesn't use ContentSettings, changes browser global state. Requires a custom handler+revoker. Revoke will likely require a browser restart.
2. ```RegisterProtocolHandlerPermissionRequest``` - registers protocol handler per site (?). Doesn't use ContentSettings, stores data in own custom backend. Requires a custom handler+revoker.
3. ```DownloadPermissionRequest``` - allows multi-download per site, uses ContentSettings, but has its own setter. Requires a custom handler, revoke will work fine.
4. ```PerDeviceProvisioningPermissionRequest``` - Android-specific, doesn't use ContentSettings, somehow connected to DRM (created from ```CreateMediaDrmStorage```). Requires a custom handler+revoker.
5. ```AttestationPermissionRequest``` - an Extension API [method](https://github.com/chromium/chromium/blob/91.0.4446.2/chrome/common/extensions/api/cryptotoken_private.idl#L50-L51) can trigger this request,  doesn't use ContentSettings. Requires a custom handler+revoker.
6. ```QuotaPermissionRequest``` - a DeprecatedStorageQuota API [method](https://github.com/chromium/chromium/blob/91.0.4450.1/third_party/blink/renderer/modules/quota/deprecated_storage_quota.idl#L33), doesn't use ContentSettings. Requires a custom handler+revoker.

* custom handler - means we need to have some code to handle the moment permission is accepted.
* custom revoker - means we need to have some code to revoke the permission.

Technically it's possible to add required logic for each of this requests, but I don't think we should do this. It's very likely that each one of them will introduce some conflict to Chromium (patches, etc), it's possible that revoke of some of them can be harmful for a runtime logic, and it's possible that some of them are very rare.

From my POV, we can/should support ```DownloadPermissionRequest``` and keep other requests as is, but maybe I don't see a full picture. @pes10k what do you think?

**Comment by pes10k** [comment]
I think that all sounds great @goodov !  And fwiw, I think `DownloadPermissionRequest` could wait till later too, as a follow up item.  Especially if making deeper changes in Chromium to support the above would make it harder to partition the permissions too (which to my mind is a higher priority).

**Comment by simonhong** [comment]
@goodov Regarding to `WidevinePermissionRequest`, I think we just ***borrowed*** permission UI to ask user whether install widevie or not globally. It would be great if we can control widevine permisison per site even if widevine plugin is installed.

**Comment by stephendonner** [comment]
@goodov is this slated for `1.23.x`, as https://github.com/brave/brave-browser/issues/14127 is?  Although we can test the UI/UX of the permissions by enabling the feature flag (which I'll do and update the issue with findings), without the core API landed, baked, and uplifted, we're not really giving it a fair shake in `1.23.x`, if so.  Thanks in advance; want to ensure we're able to give good testing coverage, which means it should land this week and be uplifted soon, if so.

/cc @rebron for timing of the above two inter-related features.

**Comment by goodov** [comment]
I think we might skip ```1.23.x```, there is no need to rush. The core implementation should land this/next week, but we also have Android UI part to implement so overall the ```1.24.x``` target is more appropriate.

**Comment by LaurenWags** [comment]
@pes10k @goodov could we get a test plan for this one please? Marking as `QA/Blocked` until we have this 👍🏻

**Comment by stephendonner** [comment]
We have a testplan in-hand (via Slack), but I'm leaving `QA/Blocked` until we have the uplift into `1.24.x` for the `until I close this window` functionality.

**Comment by kjozwiak** [comment]
Removed `QA/Blocked` as https://github.com/brave/brave-core/pull/8585 was merged. We'll need a new build but this is ready to QA once new builds are available 👍

**Comment by stephendonner** [comment]
Verified `PASSED` using 

Brave | 1.24.75 Chromium: 90.0.4430.85 (Official Build) beta (x86_64)
-- | --
Revision | 5bc145d831c180d9ff94f29a0d7a2e1cbd30ef36-refs/branch-heads/4430@{#1311}
OS | macOS Version 11.3 (Build 20E232)


NOTES (especially for the rest of the @brave/legacy_qa team):

1. keep in mind you will encounter https://github.com/brave/brave-browser/issues/15473 (and state-related variants); window and tab state aren't updated dynamically, often (I might be filing more bugs, feel free to do so as well, of course).  for shorter lifetimes, clicking on the buttons on `htttps://permissions.info` will tell you whether a permission previously granted for that site has expired or not, though.
2. not all steps are represented with screenshots, for brevity; especially, often, the final step, which is typically just a re-hash of the first step, which already has an included screenshot
3. use your best judgement re: steps and screenshots - there are multiple ways to verify the permissions' lifetimes are accurate throughout the UI (lock-icon flyout in the URL bar, `brave://settings/content`, `brave://settings/content/camera`, `brave://settings/content/microphone`, `brave://settings/content/location`, `brave://settings/content/midi`, `brave://settings/content/notifications` and more
4. some screenshots are included as part of extra verification/alternate steps you might want to take to verify -- particularly of the lock-icon flyout w/permissions, as I've tried to vary a few of the places to check (as there are more than one) the permissions
5. I purposefully also omitted the repeated posting of a screenshot for the initial step to enable `Lifetime Permissions` via `brave://flags`, since that's common to nearly all the testcases here
6. you may and are encouraged to use sites other than `https://permission.site` to test, of course!  if you use that site, be sure you are loading it over `HTTPS`, rather than `HTTP`, as many of the tests' buttons won't work over `HTTP`


## Time-based cases

**20-seconds, Allow permission** (`microphone`):
<summary>
<details>

1. launch Brave using `--enable-logging=stderr --vmodule="*/bat-native-ledger/*"=6,"*/brave_rewards/*"=6,"*/bat-native-ads/*"=6,"*/bat-native-confirmations/*"=6,"*/brave_ads/*"=9,"*/brave_user_model/*"=6 --permission-lifetime-test-seconds=20`
2. load `brave://flags` and enable the `Permissions Lifetime` flag
3. restart Brave via command-line
4. load `https://permission.site`
5. click on `Microphone`
6. Give permission for `20 seconds`, click `Allow`
7. click on the microphone icon in the URL bar
8. confirm it reads `Continue allowing permission.site to access your microphone`
9. open `brave://settings/content/microphone`
10. confirm that `https://permission.site:443` is under `Allow`
11. wait until the remainder of the original 20 seconds passes
12. on `https://permission.site`, confirm that clicking on `Microphone` again re-prompts for permissions
13. open a tab with `brave://settings/content/` and confirm there's no `microphone` or `permission.site` listed under `Recent activity` 

example | example | example | example | example
---------|----------|---------|----------|---------
<img width="1005" alt="Screen Shot 2021-04-26 at 3 05 07 PM" src="https://user-images.githubusercontent.com/387249/116157081-37bcd600-a6a1-11eb-9afa-160d7c608bf8.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 3 05 19 PM" src="https://user-images.githubusercontent.com/387249/116157082-38556c80-a6a1-11eb-86e8-5057deeb7062.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 3 05 45 PM" src="https://user-images.githubusercontent.com/387249/116157083-38556c80-a6a1-11eb-82a0-a43e63b18c32.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 3 05 51 PM" src="https://user-images.githubusercontent.com/387249/116157085-38556c80-a6a1-11eb-9313-cf41524bd66d.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 3 06 52 PM" src="https://user-images.githubusercontent.com/387249/116157088-38ee0300-a6a1-11eb-8c5a-0ea98da703eb.png">


</details>

**20-seconds, Allow permission** (`camera + microphone`):
<summary>
<details>

1. launch Brave using `--enable-logging=stderr --vmodule="*/bat-native-ledger/*"=6,"*/brave_rewards/*"=6,"*/bat-native-ads/*"=6,"*/bat-native-confirmations/*"=6,"*/brave_ads/*"=9,"*/brave_user_model/*"=6 --permission-lifetime-test-seconds=20`
2. load `brave://flags` and enable the `Permissions Lifetime` flag
3. restart Brave via command-line
4. load `https://permission.site`
5. click on `Camera + Microphone`
6. Give permission for `20 seconds`, click `Allow`
7. click on the lock icon in the URL bar
8. confirm the popup shows `Camera` - `Allow` and `Microphone` - `Allow`
9. wait the remainder of the 20 seconds
10. click on the lock icon again
11. confirm the popup no longer has `Camera` nor `Microphone` with any values present

example | example | example
---------|----------|---------
<img width="1005" alt="Screen Shot 2021-04-26 at 3 10 09 PM" src="https://user-images.githubusercontent.com/387249/116157372-b9146880-a6a1-11eb-9f3e-3e2d97ad3571.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 3 10 17 PM" src="https://user-images.githubusercontent.com/387249/116157375-b9acff00-a6a1-11eb-81d1-3c45616887df.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 3 10 36 PM" src="https://user-images.githubusercontent.com/387249/116157376-ba459580-a6a1-11eb-8259-d87df62d0c96.png">


</details>

## Origin-based cases

**Origin-based behavior, multiple tabs** (`location`):
<summary>
<details>

1. load `brave://flags` and enable the `Permission Lifetime` flag
2. restart Brave
3. load `hulu.com`
4. log in
5. Give permission `until I close this page`
6. click `Allow`
7. confirm `brave://settings/content/location` shows `Allow` for `Location`
8. click `Allow` when prompted to install Widevine
9. play a video
10. open `hulu.com` in another tab
11. play a video
12. close either of the two open tabs
13. confirm `brave://settings/content/location` shows `Allow` for `Location`
14. close the remaining tab
15. confirm `brave://settings/content/location` now shows `Ask before accessing (recommended)` and has nothing under `Block` and `Allow`

example | example | example | example | example | example
---------|----------|---------|----------|----------|--------
<img width="1005" alt="Screen Shot 2021-04-26 at 3 12 58 PM" src="https://user-images.githubusercontent.com/387249/116157886-98004780-a6a2-11eb-9e3f-b0c64cb97cdc.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 3 13 08 PM" src="https://user-images.githubusercontent.com/387249/116157891-9898de00-a6a2-11eb-9fcf-e76e84c70ebe.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 3 13 14 PM" src="https://user-images.githubusercontent.com/387249/116157892-99317480-a6a2-11eb-80f1-87cbecdcfb20.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 3 15 43 PM" src="https://user-images.githubusercontent.com/387249/116157895-99ca0b00-a6a2-11eb-951f-686dbadc9c77.png"> | <img width="1680" alt="Screen Shot 2021-04-26 at 3 16 03 PM" src="https://user-images.githubusercontent.com/387249/116157896-99ca0b00-a6a2-11eb-99be-b6f8330a1bb9.png"> | <img width="1676" alt="Screen Shot 2021-04-26 at 3 16 15 PM" src="https://user-images.githubusercontent.com/387249/116157898-9a62a180-a6a2-11eb-8029-0fb54c9f02d4.png">


</details>

**Origin-based behavior,  leave tab open, restart** (`MIDI`):
<summary>
<details>

1. load `brave://flags` and enable the `Permission Lifetime` flag
2. restart Brave
3. load `https://permission.site/`
4. click on MIDI
5. Give permission `until I close this page`
6. click `Allow`
7. confirm `brave://settings/content` shows `permission.site` with `Allowed MIDI devices`
8. leave tab open
9. restart Brave
10. confirm `brave://settings/content/midiDevices` shows `Ask when a site wants to use system exclusive messages to access MIDI MIDI devices`
11. confirm there's no sites under `Block` or `Allow`
12. load `https://permission.site/`
13. click on `MIDI`
14. confirm you now get prompted for permission

example | example | example | example | example 
---------|---------|----------|---------|----------
<img width="1005" alt="Screen Shot 2021-04-26 at 3 22 52 PM" src="https://user-images.githubusercontent.com/387249/116158564-b9adfe80-a6a3-11eb-9cef-571a998f6189.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 3 23 01 PM" src="https://user-images.githubusercontent.com/387249/116158566-ba469500-a6a3-11eb-8362-02f7b1f23163.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 3 23 08 PM" src="https://user-images.githubusercontent.com/387249/116158567-badf2b80-a6a3-11eb-84e9-efb0a5be69a7.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 3 23 17 PM" src="https://user-images.githubusercontent.com/387249/116158568-badf2b80-a6a3-11eb-9999-af2f72b257e5.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 3 23 55 PM" src="https://user-images.githubusercontent.com/387249/116158569-bb77c200-a6a3-11eb-9705-231b9b56807b.png">


</details>

**Origin-based behavior,  close tab, open new tab** (`MIDI`):
<summary>
<details>

1. load `brave://flags` and enable the `Permission Lifetime` flag
2. restart Brave
3. load `https://permission.site/`
4. click on `MIDI`
5. Give permission `until I close this page`
6. click `Allow`
7. confirm `brave://settings/content` shows `permission.site` with `Allowed MIDI devices`
8. close `permission.site` tab
9. open `brave://settings/content/midiDevices` in a new tab
10. confirm there are no permissions or sites listed under `Block` or `Allow`
11. open `https://permission.site/` in a new tab
12. click on `MIDI`
13. confirm you get prompted for permission 

example | example | example
---------|---------|----------
<img width="1005" alt="Screen Shot 2021-04-26 at 3 30 54 PM" src="https://user-images.githubusercontent.com/387249/116159293-d39c1100-a6a4-11eb-998f-f115b57f6500.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 3 31 06 PM" src="https://user-images.githubusercontent.com/387249/116159294-d434a780-a6a4-11eb-8c2f-1133c6a9a6b8.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 3 33 04 PM" src="https://user-images.githubusercontent.com/387249/116159296-d4cd3e00-a6a4-11eb-99b9-c3c727e75199.png">


</details>

## 24-hour / 1-week cases

**25-hours, negative test** (`microphone`):
<summary>
<details>

1. load `brave://flags` and enable the `Permission Lifetime` flag
2. restart Brave
3. load `https://permission.site/`
4. click on `Microphone`
5. give permission `for 24 hours`
6. confirm `brave://settings/content` shows `Allow` for `Microphone`
7. shut down Brave
8. set system clock forward by 25 hours
9. confirm `brave://settings/content` shows `Ask (default)` for `Microphone`
10. load `https://permission.site/`
11. confirm you get re-prompted for microphone permissions when clicking on `Microphone`

example | example | example | example
---------|---------|----------|----------
<img width="900" alt="Screen Shot 2021-04-26 at 3 40 02 PM" src="https://user-images.githubusercontent.com/387249/116160494-00512800-a6a7-11eb-9ea6-139cabf1674f.png"> | <img width="902" alt="Screen Shot 2021-04-26 at 3 37 33 PM" src="https://user-images.githubusercontent.com/387249/116160528-10690780-a6a7-11eb-9d02-c01354d4b6dc.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 3 37 52 PM" src="https://user-images.githubusercontent.com/387249/116160529-11019e00-a6a7-11eb-8a7b-2f1208662d8c.png"> | <img width="839" alt="Screen Shot 2021-04-26 at 3 43 22 PM" src="https://user-images.githubusercontent.com/387249/116160530-119a3480-a6a7-11eb-87b8-6ac2a6a8e84c.png">





</details>

**23-hours positive test** (`microphone`):
<summary>
<details>

1. load `brave://flags` and enable the `Permission Lifetime` flag
2. restart Brave
3. load `https://permission.site/`
4. click on `Microphone`
5. give permission `for 24 hours`
6. confirm `brave://settings/content` shows `Allow` for `Microphone`
7. shut down Brave
8. set system clock forward by 25 hours
9. confirm `brave://settings/content` shows `Ask (default)` for `Microphone`
10. load `https://permission.site/`
11. confirm you do NOT get re-prompted for microphone permissions when clicking on `Microphone`

example | example | example | example
---------|---------|----------|---------
<img width="807" alt="Screen Shot 2021-04-27 at 2 44 21 PM" src="https://user-images.githubusercontent.com/387249/116160371-c2540400-a6a6-11eb-9bac-e207fe586afa.png"> | <img width="839" alt="Screen Shot 2021-04-27 at 2 44 29 PM" src="https://user-images.githubusercontent.com/387249/116160373-c2ec9a80-a6a6-11eb-87bc-32808b958a06.png"> | <img width="839" alt="Screen Shot 2021-04-27 at 2 44 43 PM" src="https://user-images.githubusercontent.com/387249/116160376-c3853100-a6a6-11eb-97f1-d3006b978b8f.png"> | <img width="839" alt="Screen Shot 2021-04-27 at 2 45 17 PM" src="https://user-images.githubusercontent.com/387249/116160379-c3853100-a6a6-11eb-821f-d2d4db8835f7.png">



</details>

**1-week, negative test** (`camera`):
<summary>
<details>

1. load `brave://flags` and enable the `Permission Lifetime` flag
2. restart Brave
3. load `https://permission.site/`
4. click on `Camera`
5. give permission `for 1 week`
6. confirm `brave://settings/content` shows `Allow` for `Camera`
7. shut down Brave
8. set system clock forward by 8 days
9. confirm `brave://settings/content shows "Ask (default)` for `Camera`
10. load `https://permission.site/`
11. confirm you get re-prompted for microphone permissions when clicking on `Camera`

example | example | example | example
---------|----------|---------|----------
<img width="1086" alt="Screen Shot 2021-04-26 at 4 04 49 PM" src="https://user-images.githubusercontent.com/387249/116162056-dbaa7f80-a6a9-11eb-8490-b698c0e609c6.png"> | <img width="770" alt="Screen Shot 2021-04-26 at 4 07 43 PM" src="https://user-images.githubusercontent.com/387249/116162044-d3eadb00-a6a9-11eb-9de7-bf8d1a3bf9d7.png"> | <img width="1312" alt="Screen Shot 2021-05-04 at 4 07 48 PM" src="https://user-images.githubusercontent.com/387249/116162096-f7158a80-a6a9-11eb-8e22-a31955ab4768.png"> |<img width="1312" alt="Screen Shot 2021-05-04 at 4 07 58 PM" src="https://user-images.githubusercontent.com/387249/116162100-f7ae2100-a6a9-11eb-86b1-0d1338be4413.png">






</details>

**1-week, positive test** (`location`):
<summary>
<details>

1. load `brave://flags` and enable the `Permission Lifetime` flag
2. restart Brave
3. load `https://permission.site/`
4. click on `Location`
5. give permission `for 1 week`
6. confirm `brave://settings/content` shows `Allow` for `Location`
7. shut down Brave
8. set system clock forward by 6 days
9. confirm `brave://settings/content shows `permission.site` with `Allowed location` under `Recent activity`
10. load `https://permission.site/`
11. confirm you do NOT get re-prompted for location permissions when clicking on `Location`

example | example | example | example | example
---------|----------|---------|----------|---------
<img width="1086" alt="Screen Shot 2021-04-26 at 3 55 30 PM" src="https://user-images.githubusercontent.com/387249/116161361-7dc96800-a6a8-11eb-990a-fb34c303eca2.png"> | <img width="1086" alt="Screen Shot 2021-04-26 at 3 55 42 PM" src="https://user-images.githubusercontent.com/387249/116161366-7e61fe80-a6a8-11eb-85b7-6cbc91ff364f.png"> |<img width="992" alt="Screen Shot 2021-04-26 at 3 56 21 PM" src="https://user-images.githubusercontent.com/387249/116161368-7efa9500-a6a8-11eb-94ff-ed0be1eb83ea.png"> |  <img width="995" alt="Screen Shot 2021-05-02 at 3 56 38 PM" src="https://user-images.githubusercontent.com/387249/116161412-933e9200-a6a8-11eb-80c8-824bb692f424.png"> | <img width="1086" alt="Screen Shot 2021-05-02 at 3 57 29 PM" src="https://user-images.githubusercontent.com/387249/116161414-93d72880-a6a8-11eb-8e41-f62d783819ee.png">





</details>

## Allow/Block forever cases

**Allow forever** (`location`):
<summary>
<details>

1. load `brave://flags` and enable the `Permission Lifetime` flag
2. restart Brave
3. load `https://permission.site/`
4. click on `Location`
5. give permission `forever`
6. confirm `brave://settings/content` shows `Allow` for `Location`
7. shut down Brave
8. set system clock forward by 2 months
9. confirm `brave://settings/content` shows `Allow` for `Location`
10. load `https://permission.site/`
11. confirm you get do NOT get reprompted for Location

example | example | example | example | example
---------|---------|----------|----------|---------
<img width="1312" alt="Screen Shot 2021-04-26 at 4 20 21 PM" src="https://user-images.githubusercontent.com/387249/116163384-8328b180-a6ac-11eb-9b12-fdb3b38133eb.png"> | <img width="1312" alt="Screen Shot 2021-04-26 at 4 20 43 PM" src="https://user-images.githubusercontent.com/387249/116163386-84f27500-a6ac-11eb-9d46-50af94946e34.png"> | <img width="675" alt="Screen Shot 2021-04-26 at 4 21 15 PM" src="https://user-images.githubusercontent.com/387249/116163437-9dfb2600-a6ac-11eb-85a7-679a3e024b16.png"> | <img width="1216" alt="Screen Shot 2021-06-26 at 4 22 43 PM" src="https://user-images.githubusercontent.com/387249/116163465-aa7f7e80-a6ac-11eb-89ef-c6ae527684b3.png"> | <img width="1203" alt="Screen Shot 2021-06-26 at 4 22 50 PM" src="https://user-images.githubusercontent.com/387249/116163468-ab181500-a6ac-11eb-9d79-5d2fa225f8ae.png">

</details>

**Block forever** (`notifications`):
<summary>
<details>

1. load `brave://flags` and enable the `Permission Lifetime` flag
2. restart Brave
3. load `https://permission.site/`
4. click on `Notifications`
5. click on `Block forever`
6. confirm `brave://settings/content` shows `Block` for `Notifications`
7. shut down Brave
8. set system clock forward by 3 months
9. confirm `brave://settings/content` shows `Block` for Notifications
10. load `https://permission.site/`
11. confirm you get do NOT get reprompted for Notifications

example | example | example | example
---------|----------|---------|----------
<img width="1312" alt="Screen Shot 2021-04-26 at 4 34 13 PM" src="https://user-images.githubusercontent.com/387249/116164132-0991c300-a6ae-11eb-8b66-8e2382c9e79c.png"> | <img width="1010" alt="Screen Shot 2021-06-26 at 4 37 58 PM" src="https://user-images.githubusercontent.com/387249/116164072-e830d700-a6ad-11eb-9f34-c27a6b395942.png"> | <img width="682" alt="Screen Shot 2021-04-26 at 4 34 53 PM" src="https://user-images.githubusercontent.com/387249/116164166-1dd5c000-a6ae-11eb-8580-e0545e96ec5a.png"> | <img width="1010" alt="Screen Shot 2021-06-26 at 4 35 29 PM" src="https://user-images.githubusercontent.com/387249/116164187-29c18200-a6ae-11eb-8b6d-dc6746138c80.png">

</details>

## Consecutive-revocation cases

**Consecutive-revocation permissions** (`location`):
<summary>
<details>

1. launch Brave using `--enable-logging=stderr --vmodule="*/bat-native-ledger/*"=6,"*/brave_rewards/*"=6,"*/bat-native-ads/*"=6,"*/bat-native-confirmations/*"=6,"*/brave_ads/*"=9,"*/brave_user_model/*"=6 --permission-lifetime-test-seconds=20`
2. load `brave://flags` and enable the `Permissions Lifetime` flag
3. restart Brave via command-line
4. load `https://permission.site`
5. click on `Location`
6. give permission for `20 seconds`
7. click on the lock icon next to `permission.site` in the URL bar
8. confirm `Location` is set to `Allow`
9. wait 10 seconds
10. open `https://browserleaks.com/geo`
11. give permission for 20 seconds
12. return to `permissions.site`
13. click on the lock icon; confirm you no longer see `Location` with any value
14. return to `browserleaks.com/geo`
15. click on the lock icon; confirm you no longer see `Location` with any value

example | example | example | example | example
---------|----------|---------|----------|---------
<img width="1005" alt="Screen Shot 2021-04-26 at 2 56 02 PM" src="https://user-images.githubusercontent.com/387249/116155980-bf094a00-a69f-11eb-9e51-8795abfd90dd.png">| <img width="1005" alt="Screen Shot 2021-04-26 at 2 56 07 PM" src="https://user-images.githubusercontent.com/387249/116155984-bfa1e080-a69f-11eb-8c03-d654b00e67f0.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 2 56 27 PM" src="https://user-images.githubusercontent.com/387249/116155986-c03a7700-a69f-11eb-8aab-f4c8db23eecf.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 2 56 41 PM" src="https://user-images.githubusercontent.com/387249/116155989-c03a7700-a69f-11eb-9ad9-5dfeb56a5209.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 2 56 52 PM" src="https://user-images.githubusercontent.com/387249/116155990-c0d30d80-a69f-11eb-9a2a-600af2a31bc9.png"> 


</details>

## Negative/current-parity cases

**Block** (`Auto Download`):
<summary>
<details>

1. launch Brave using `--enable-logging=stderr --vmodule="*/bat-native-ledger/*"=6,"*/brave_rewards/*"=6,"*/bat-native-ads/*"=6,"*/bat-native-confirmations/*"=6,"*/brave_ads/*"=9,"*/brave_user_model/*"=6 --permission-lifetime-test-seconds=20`
2. load `brave://flags` and enable the `Permissions Lifetime` flag.  restart Brave via command-line
3. load `https://permission.site`
4. click on `Auto Download`
5. verify a file picker is invoked, as well as a permissions request to `Download multiple files` with `Block` and `Allow` buttons
6. click on `Block` in the dialog
7. click `Cancel` in the file picker
8. click on the `Download` icon in the URL bar
9. confirm you see `Continue blocking automatic downloads of multiple files` chosen
10. load `brave://settings/content/automaticDownloads` and confirm `https://permission.site:443` is under the `Block` section
11. load `brave://settings/content/` and confirm `permission.site - http` reads `Blocked automatic downloads`
12. on `permission.site` click again on `Auto Download`, and confirm nothing happens (no file picker or permissions dialog pop up)

example | example | example
---------|---------|----------
<img width="1005" alt="Screen Shot 2021-04-26 at 2 53 02 PM" src="https://user-images.githubusercontent.com/387249/116155671-43a79880-a69f-11eb-8714-0f31d1e7d838.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 2 53 11 PM" src="https://user-images.githubusercontent.com/387249/116155676-44402f00-a69f-11eb-8e8c-cfe4c456739a.png"> | <img width="1005" alt="Screen Shot 2021-04-26 at 2 53 15 PM" src="https://user-images.githubusercontent.com/387249/116155679-44d8c580-a69f-11eb-9561-36800be51998.png">


</details>

**Allow** (`Auto Download`):
<summary>
<details>

1. launch Brave using `--enable-logging=stderr --vmodule="*/bat-native-ledger/*"=6,"*/brave_rewards/*"=6,"*/bat-native-ads/*"=6,"*/bat-native-confirmations/*"=6,"*/brave_ads/*"=9,"*/brave_user_model/*"=6 --permission-lifetime-test-seconds=20`
2. load `brave://flags` and enable the `Permissions Lifetime` flag.  restart Brave via command-line
3. load `https://permission.site`
4. click on `Auto Download`
5. verify a file picker is invoked, as well as a permissions request to `Download multiple files` with `Block` and `Allow` buttons
6. click on `Allow` in the dialog
7. click either `Cancel` or `Save` in the file picker
8. confirm you get prompted another time, and choose either `Cancel` or `Save` again
9. click on the `Download` icon in the URL bar
10. confirm you see `Continue allowing automatic downloads of multiple files` chosen
11. load `brave://settings/content/automaticDownloads` and confirm `https://permission.site:443` is under the `Allow` section
12. load `brave://settings/content/` and confirm `permission.site - http` reads `Allowed automatic downloads`
13. on `permission.site` click again on `Auto Download`, and confirm you again get two consecutive file-picker dialogs

example | example | example | example
---------|---------|----------|---------
<img width="942" alt="Screen Shot 2021-04-26 at 2 49 59 PM" src="https://user-images.githubusercontent.com/387249/116155339-d3007c00-a69e-11eb-99b3-f7aae8516f07.png"> | <img width="942" alt="Screen Shot 2021-04-26 at 2 50 10 PM" src="https://user-images.githubusercontent.com/387249/116155345-d3991280-a69e-11eb-844e-0b8a8c713d24.png"> | <img width="942" alt="Screen Shot 2021-04-26 at 2 50 14 PM" src="https://user-images.githubusercontent.com/387249/116155347-d3991280-a69e-11eb-80ba-162b130e3544.png"> | <img width="942" alt="Screen Shot 2021-04-26 at 2 50 21 PM" src="https://user-images.githubusercontent.com/387249/116155349-d431a900-a69e-11eb-8062-550d5e5dd5bc.png">


</details>

**Allow** (`Widevine`):
<summary>
<details>

1. new profile
2. launch Brave using `--enable-logging=stderr --vmodule="*/bat-native-ledger/*"=6,"*/brave_rewards/*"=6,"*/bat-native-ads/*"=6,"*/bat-native-confirmations/*"=6,"*/brave_ads/*"=9,"*/brave_user_model/*"=6 --permission-lifetime-test-seconds=20`
3. load `brave://flags` and enable the `Permissions Lifetime` flag.  restart Brave via command-line
4. load `netflix.com` (or any `Widevine`-using site)
5. log in
6. choose any value for `Notifications`, when prompted (or just dismiss the dialog)
7. confirm you see an `Install and run Widevine` prompt, with a `[ ] Don't ask again` checkbox and `Block` and `Allow` buttons
8. click `Allow`
9. confirm you can play a media title
10. relaunch Brave
11. play any title and confirm you are not re-prompted for `Widevine`

example | example | example
---------|----------|---------
<img width="1312" alt="Screen Shot 2021-04-26 at 10 12 58 PM" src="https://user-images.githubusercontent.com/387249/116188801-d87cb700-a6dc-11eb-9a91-814db898c55e.png"> | <img width="1312" alt="Screen Shot 2021-04-26 at 10 13 35 PM" src="https://user-images.githubusercontent.com/387249/116188804-d9154d80-a6dc-11eb-8300-ecb6ea6e81f3.png"> | <img width="1312" alt="Screen Shot 2021-04-26 at 10 14 37 PM" src="https://user-images.githubusercontent.com/387249/116188806-d9ade400-a6dc-11eb-8337-3c10d0133115.png">


</details>

## Migration (stored prefs)

**Block** (`Notifications`):
<summary>
<details>

1. using release build `1.23.73`, load `permission.site` in a new profile
2. click on `Notifications`
3. click `Block`
4. shut down
5. copy your profile over and launch `1.24.x` with it
6. load `brave://flags` and flip `Permission Lifetime` to `Enabled`
7. restart Brave `1.24.x`
8. load `permission.site`
9. confirm that clicking on `Notifications` doesn't prompt or show notifications
10. load `brave://settings/content/notifications` and confirm that `https://permission.site:443` is listed under `Block`

example | example
---------|---------
<img width="997" alt="Screen Shot 2021-04-26 at 2 44 20 PM" src="https://user-images.githubusercontent.com/387249/116154769-f5de6080-a69d-11eb-9653-e529e987f576.png"> | <img width="997" alt="Screen Shot 2021-04-26 at 2 40 20 PM" src="https://user-images.githubusercontent.com/387249/116154492-954f2380-a69d-11eb-8ce8-6f158b2b8b67.png">


</details>

**Allow** (`Location`):
<summary>
<details>

1. using Brave release build `1.23.73`, load `permission.site` in a new profile
2. click on Location
3. click `Allow`
4. shut down
5. copy your profile over and launch `1.24.x` with it
6. load `brave://flags` and flip `Permission Lifetime` to `Enabled`
7. restart Brave `1.24.x`
8. load `permission.site`
9. confirm that clicking on `Location` doesn't prompt or show notifications
10. load `brave://settings/content/location` and confirm that `https://permission.site:443` is listed under `Allow`
11. click on the dropped-pin icon in the URL bar and confirm it says `Continue allowing this site to access your location`


example | example
---------|---------
<img width="997" alt="Screen Shot 2021-04-26 at 2 36 59 PM" src="https://user-images.githubusercontent.com/387249/116154014-0215ee00-a69d-11eb-9e46-464194867400.png"> | <img width="997" alt="Screen Shot 2021-04-26 at 2 37 13 PM" src="https://user-images.githubusercontent.com/387249/116154015-02ae8480-a69d-11eb-94d6-9cada4a8b696.png">
</details>

**Comment by srirambv** [comment]
<details>
<summary>Verification passed on OnePlus 6T with Android 10 running 1.24.81 x64 build</summary>

### Time Based checks

<details>
<summary>20 Sec timer - Allow (Camera + Microphone)</summary>
<br>

- Verified passing argument to set permission timer to 20 sec shows up as an option
- Verified selecting 20 sec option and allow enables the Camera + Microphone on the site and system notification shows up
- Verified site is added under allow list for both Camera + Microphone
- Verified after 20 sec the site permissions are removed automatically
- Verified reloading the page and selecting Camera on the page triggers the permission prompt again

20 Sec option | Select 20 sec timer | Site requesting permission | Site added to allow list 
-- | -- | -- | --
![image](https://user-images.githubusercontent.com/17010094/116682083-b6698a00-a9a5-11eb-9eaf-c215404c44e8.png) | ![image](https://user-images.githubusercontent.com/17010094/116682096-b9fd1100-a9a5-11eb-84ed-923ca8e0dc89.png) | ![image](https://user-images.githubusercontent.com/17010094/116682187-d436ef00-a9a5-11eb-9725-df4f5906a61c.png) | ![image](https://user-images.githubusercontent.com/17010094/116682161-cbdeb400-a9a5-11eb-866c-11156c2a9bb8.png)


</details>

<details>
<summary>20 Sec timer - Block Forever (Camera + Microphone)</summary>
<br>

- Verified setting permission to `Block Forever` adds an entry under blocked list for the site
  <img width=300 src=https://user-images.githubusercontent.com/17010094/116681675-304d4380-a9a5-11eb-8401-4fd4233e30e5.png>

</details>
<hr>

### Origin-based cases

<details>
<summary>Origin-based behaviour, multiple tabs (Location)</summary>
<br>

- Verified visiting `browserleaks.com/geo` triggers permission prompt
- Verified selecting `Until I close this page` and select allow sets the permission for the page and detects location
- Verified visiting permission.site on a second tab and selecting location triggers prompt
- Verified selecting `Block Forever` adds the site to blocked list
- Verified site settings shows both sites under allowed/blocked respectively
- Verified closing `browserleaks.com/geo` tab resets the site permission from site settings
- Verified revisiting the page on a new tab triggers the prompt again

</details>

<hr>

### 24-hour / 1-week cases

<details>
<summary>23 hrs test </summary>
<br>

- Verified allowing permission on a site for 24hrs works fine
- Verified restarting the browser and loading the page doesn't trigger permission prompt
- Verified moving device date ahead to 23hrs and reloading the page/opening in a new tab doesn't trigger the permission prompt

</details>
<details>
<summary>25 hrs test </summary>
<br>

- Verified allowing permission on a site for 24hrs works fine
- Verified restarting the browser and loading the page doesn't trigger permission prompt
- Verified moving device date ahead to 25hrs and reloading the page/opening in a new tab trigger the permission prompt

</details>
<details>
<summary>6 Days 12hrs test </summary>
<br>

- Verified allowing permission on a site for 1 week works fine
- Verified restarting the browser and loading the page doesn't trigger permission prompt
- Verified moving device date ahead to 6 days & 12 hrs ahead and reloading the page/opening in a new tab doesn't trigger the permission prompt

</details>
<details>
<summary>8 days Test</summary>
<br>

- Verified allowing permission on a site for 1 week works fine
- Verified restarting the browser and loading the page doesn't trigger permission prompt
- Verified moving device date ahead to 8days and reloading the page/opening in a new tab trigger the permission prompt

</details>

### Allow/Block forever cases
<hr />
<details>
<summary>Allow forever (Location)</summary>
<br>

- Verified selecting location on `https://permission.site` triggers permission prompt
- Verified able to set the permission to `forever`
- Verified reloading the page doesn't trigger the prompt
- Verified site settings adds the site to allow list
- Verified moving device date ahead by a month and revisiting the page in a new tab doesn't trigger the notification

</details>
<details>
<summary>Block Forever (Notifications)</summary>
<br>

- Verified selecting location on `https://permission.site` triggers permission prompt
- Verified able to set the permission to `Block Forever`
- Verified site settings adds the site to block list 
- Verified moving device date ahead by a month and revisiting the page in a new tab doesn't trigger the notification

</details>

<hr>

### Consecutive-revocation cases

<details>
<summary>10 sec repeat permission</summary>
<br>

- Verified launching with command line `--permission-lifetime-test-seconds=10` shows 10 sec option in permission prompt
- Verified setting the prompt for 10 sec on `https://permission.site` works
- Opening a new tab and visiting `https://browserleaks.com/geo` triggers the prompt and set it to 10 sec option
- Wait for 10sec and navigate back to permission site via bottom toolbar swipe prompts again to set permission
- Swipe back to browserleaks tab prompts on that page as well

</details>


### Private tab test

<details>
<summary>Permission on private tabs</summary>
<br>

- Verified permission prompt shows up on private tab as well
- Verified setting persmisson forever works on private tab
- Verified closing the private tab and opening a new one and visiting the same page triggers the permission prompt
- Verified during an active session, setting permisson on private tab is only valid for the session
- Verified during an active session, setting permission forever and opening the same page on normal tab still prompts for permission setting

</details>

- Verified clearing site setting via clear data resets all permissions set on any page

</details>

----------------------------

<details>
<summary>Verification passed on Samsung Tab A with Android 10 running 1.24.81 x64 build</summary>

### Time Based checks

<details>
<summary>20 Sec timer - Allow (Camera + Microphone)</summary>
<br>

- Verified passing argument to set permission timer to 20 sec shows up as an option
- Verified selecting 20 sec option and allow enables the Camera + Microphone on the site and system notification shows up
- Verified site is added under allow list for both Camera + Microphone
- Verified after 20 sec the site permissions are removed automatically
- Verified reloading the page and selecting Camera on the page triggers the permission prompt again

20 Sec option | Select 20 sec timer | Site requesting permission | Site added to allow list 
-- | -- | -- | --


</details>

<details>
<summary>20 Sec timer - Block Forever (Camera + Microphone)</summary>
<br>

- Verified setting permission to `Block Forever` adds an entry under blocked list for the site
  <img width=300 src=>

</details>

<hr>

### Origin-based cases

<details>
<summary>Origin-based behaviour, multiple tabs (Location)</summary>
<br>

- Verified visiting `browserleaks.com/geo` triggers permission prompt
- Verified selecting `Until I close this page` and select allow sets the permission for the page and detects location
- Verified visiting permission.site on a second tab and selecting location triggers prompt
- Verified selecting `Block Forever` adds the site to blocked list
- Verified site settings shows both sites under allowed/blocked respectively
- Verified closing `browserleaks.com/geo` tab resets the site permission from site settings
- Verified revisiting the page on a new tab triggers the prompt again

</details>

<hr>

### 24-hour / 1-week cases

<details>
<summary>23 hrs test </summary>
<br>

- Verified allowing permission on a site for 24hrs works fine
- Verified restarting the browser and loading the page doesn't trigger permission prompt
- Verified moving device date ahead to 23hrs and reloading the page/opening in a new tab doesn't trigger the permission prompt

</details>
<details>
<summary>25 hrs test </summary>
<br>

- Verified allowing permission on a site for 24hrs works fine
- Verified restarting the browser and loading the page doesn't trigger permission prompt
- Verified moving device date ahead to 25hrs and reloading the page/opening in a new tab trigger the permission prompt

</details>
<details>
<summary>6 Days 12hrs test </summary>
<br>

- Verified allowing permission on a site for 1 week works fine
- Verified restarting the browser and loading the page doesn't trigger permission prompt
- Verified moving device date ahead to 6 days & 12 hrs ahead and reloading the page/opening in a new tab doesn't trigger the permission prompt

</details>
<details>
<summary>8 days Test</summary>
<br>

- Verified allowing permission on a site for 1 week works fine
- Verified restarting the browser and loading the page doesn't trigger permission prompt
- Verified moving device date ahead to 8days and reloading the page/opening in a new tab trigger the permission prompt

</details>

### Allow/Block forever cases
<hr />
<details>
<summary>Allow forever (Location)</summary>
<br>

- Verified selecting location on `https://permission.site` triggers permission prompt
- Verified able to set the permission to `forever`
- Verified reloading the page doesn't trigger the prompt
- Verified site settings adds the site to allow list
- Verified moving device date ahead by a month and revisiting the page in a new tab doesn't trigger the notification

</details>
<details>
<summary>Block Forever (Notifications)</summary>
<br>

- Verified selecting location on `https://permission.site` triggers permission prompt
- Verified able to set the permission to `Block Forever`
- Verified site settings adds the site to block list 
- Verified moving device date ahead by a month and revisiting the page in a new tab doesn't trigger the notification

</details>

<hr>

### Consecutive-revocation cases

<details>
<summary>10 sec repeat permission</summary>
<br>

- Verified launching with command line `--permission-lifetime-test-seconds=10` shows 10 sec option in permission prompt
- Verified setting the prompt for 10 sec on `https://permission.site` works
- Opening a new tab and visiting `https://browserleaks.com/geo` triggers the prompt and set it to 10 sec option
- Wait for 10sec and navigate back to permission site via bottom toolbar swipe prompts again to set permission
- Swipe back to browserleaks tab prompts on that page as well

</details>


### Private tab test

<details>
<summary>Permission on private tabs</summary>
<br>

- Verified permission prompt shows up on private tab as well
- Verified setting persmisson forever works on private tab
- Verified closing the private tab and opening a new one and visiting the same page triggers the permission prompt
- Verified during an active session, setting permisson on private tab is only valid for the session
- Verified during an active session, setting permission forever and opening the same page on normal tab still prompts for permission setting

</details>

- Verified clearing site setting via clear data resets all permissions set on any page

</details>
